### DISCUSSION TITLE
Retrieve full bill text from the 'https://v3.openstates.org/bills' endpoint

### DISCUSSION THREAD

**Discussion by byronpoplawski** [openstates/issues#1038] [Q&A] [answered]
Hey is there a way to retrieve the full text associated with a bill? 

I love the API and I can successfully pull a list of bills and see their titles/metadata but for the project I'm working on, I want to pull the full bill text in in HTML or some other format. Is there a way to do that with the /bills endpoint?

**Comment by Fowlplaychiken** [comment]
Same here. Also for a projectOn May 12, 2023, at 1:12 AM, byronpoplawski ***@***.***> wrote:﻿
Hey is there a way to retrieve the full text associated with a bill?
I love the API and I can successfully pull a list of bills and see their titles/metadata but for the project I'm working on, I want to pull the full bill text in in HTML or some other format. Is there a way to do that with the /bills endpoint?

**Comment by NewAgeAirbender** [comment, marked as answer]
No, there is not a way to pull bill text through the API. If you created a page scraper to pull text using the version links provided through the API, that would be the best way to get it.

**Reply by byronpoplawski** [reply]
Ok, do you happen to know whether there are any other free API services that do enable you to pull a full bill text?

**Reply by NewAgeAirbender** [reply]
Depending if the jurisdiction has an API for their site, sometimes they have bill text available there. I don't know of any states that have it for sure off the top of my head though.

**Reply by byronpoplawski** [reply]
Got it thanks! I just looked into your scraping method and that probably will work too :) Thanks! I hope open states doesn't mind me scraping the bill text.

**Reply by byronpoplawski** [reply]
Out of curiosity, do you have any preferred Python libraries for web scraping like this?

**Reply by NewAgeAirbender** [reply]
Our founder wrote a nifty one called [Spatula](https://jamesturk.github.io/spatula/) which we use to get most of our People & Committee data. It's pretty easy to get the hang of & very readable!

**Reply by byronpoplawski** [reply]
Thanks!

**Comment by Fowlplaychiken** [comment]
None I have found. I need a way to pull full bill and statute text, preferably in plaintext with font characteristics represented as html. Ive resigned myself to needing to create a data scraper.Working on a GPT program to analyze, and draft, legislation.On May 12, 2023, at 2:30 PM, byronpoplawski ***@***.***> wrote:﻿
Ok, do you happen to know whether there are any other free API services that do enable you to pull a full bill text?

**Reply by byronpoplawski** [reply]
Lol I'm building a similar thing :) Mostly just to analyze the regulation, not to draft it. Great minds think alike.

**Reply by byronpoplawski** [reply]
If you end up building a good scraper let me know! I'd love to compare code. I'll do the same

**Comment by Fowlplaychiken** [comment]
We should collab 🤣Im setting it to generate press releases, media advisories, cosponsorship memos, bill analysis, new legislation modifying existing statute, amendments modifying existing legislation. May train it on some legal writing tasks as well. I figure if it works, Lexus Nexus or Lockheed Martins Internet Quorum team might have an interest. I know those guys; this sort of functionality is a step above what Statenet or Quorum offers.On May 12, 2023, at 2:36 PM, byronpoplawski ***@***.***> wrote:﻿
Lol I'm building a similar thing :) Mostly just to analyze the regulation, not to draft it. Great minds think alike.

**Reply by byronpoplawski** [reply]
Ah very cool. My goals are less ambitious. I’m creating a script that pulls the latest congressional bills introduced in the WA legislature and then passes the text to chatGPT to get a summary of the bill and the pros/critiques. Then it posts it on Twitter like a Twitter bot as a way of helping regular people keep up to date on what’s happening in the WA government.

**Reply by ByronPop** [reply]
Hey following up here, I created the scraper and the Twitter bot. I created a ReadME doc that explains the project: https://github.com/ByronPop/BillBot.

If you're interested in seeing my most recent state, check out https://twitter.com/WABillBot

**Comment by Fowlplaychiken** [comment]
Likewise, will do!On May 12, 2023, at 2:37 PM, byronpoplawski ***@***.***> wrote:﻿
If you end up building a good scraper let me know! I'd love to compare code. I'll do the same

**Comment by Fowlplaychiken** [comment]
Very neat. Mine is more for us internally in the state or Federal legislature. Still similar. Probably requires a lot of the same functionality.Good luck! Very curious to see what you come up with!On May 12, 2023, at 3:21 PM, byronpoplawski ***@***.***> wrote:﻿
Ah very cool. My goals are less ambitious. I’m creating a script that pulls the latest congressional bills introduced in the WA legislature and then passes the text to chatGPT to get a summary of the bill and the pros/critiques. Then it posts it on Twitter like a Twitter bot as a way of helping regular people keep up to date on what’s happening in the WA government.
